### ISSUE TITLE
OpenApi v3 Structural Error for extensionProperties

### ISSUE THREAD

**Issue by frost-byte** [api-platform/core#4087]
**API Platform version(s) affected**: 2.6.2

**Description**  
My Swagger documentation is not including any routes.
When I take the contents of /docs.json?spec_version=3 and paste them into editor.swagger.io
it shows a lot of errors. (I tell it to convert the configuration to yaml)

One in particular is `Structural error at should NOT have additional properties additionalProperty: extensionProperties`
There are many keys for `extensionProperties: []`, is this an additional feature added to the spec by api platform?
Is there a way to configure the open api documentation to exclude entries for keys that are null or empty objects?

**Additional Context**  
Removing all lines that include extensionProperties obviously removes any warnings about this.
The same goes for any properties that have been configured as [] or null

**Comment by soyuka** [comment]
Mhh `extensionsProperties` is an internal field that when normalized outputs `x-` keys which are extension points [as documented here](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#specificationExtensions). If this key show up it's a bug, may I have more informations?

**Comment by frost-byte** [comment]
Sure, what would you like me to provide?  I can include the header and the routes for an entity as they appear during an export on the commandline or from using the endpoint.  I disabled my custom normalizers and cleared the cache, but that didn’t make a difference.  I’m using Symfony 5.2 and php 7.4.3.  I’ve made a few adjustments to api-platform locally to see if I could adjust the paths, which are cone out nested inside of an additional “paths” key.  When I do that my docs show endpoints again in swagger (but I can’t use the docs page to send queries). I can call my entity endpoints from postman in various formats: json, json+ld and they seem to work.

I do have some issues with some circular references that I couldn’t address with annotations, I was able to create a work around to convert some of the nested entities to iri’s, but I don’t think I’d normally have to do that. (Grouping should handle that)

My only guess is that the normalization process for OpenApi must not be performing all of the steps, including processing the extensionProperties

**Comment by frost-byte** [comment]
Here's some of my configuration.
I've tried with v2.6.2 and dev-master, it doesn't seem to make a difference.
I've also cleared the cache, but it doesn't seem to make a difference. (even deleting the project's var/cache/dev/ contents)

`config/packages/api_platform.yaml`

```yaml
api_platform:
    title: 'My Api'
    description: 'My Api description'
    version: '0.1.0'
    openapi:
        contact:
            name: 'My Name'
            email: 'me@email.com'
    swagger:
        versions: [3]
    mapping:
        paths: ['%kernel.project_dir%/src/Entity']
    formats:
        jsonld:
            mime_types:
                - application/ld+json
        json:
            mime_types:
                - application/json
        html:
            mime_types:
                - text/html
        jsonhal:
            mime_types:
                - application/hal+json
        csv:
            mime_types:
                - text/csv
        jsonapi:
            mime_types:
                - application/vnd.api+json

```
```yaml
# config/services.yaml

services:
    _defaults:
        autowire: true      # Automatically injects dependencies in your services.
        autoconfigure: true # Automatically registers your services as commands, event subscribers, etc.

    App\:
        resource: '../src/'
        exclude:
            - '../src/DependencyInjection/'
            - '../src/Entity/'
            - '../src/Migrations/'
            - '../src/Kernel.php'
            - '../src/Tests/'

    App\Controller\:
        resource: '../src/Controller/'
        tags: ['controller.service_arguments']
        
    # Custom Services (commenting them out and changing the service class files to non-php, i.e. ClassName.phpbk and clearing
    # the cache using bin/console cache:clear does not change anything.
    App\Serializer\Normalizer\DateNormalizer:
        tags:
            - { name: 'serializer.normalizer', priority: 64 }

    App\Serializer\Normalizer\MetricNormalizer:
        autowire: false
        autoconfigure: true
        arguments:
            $normalizer: '@api_platform.jsonld.normalizer.item'
            $iriConverter: '@ApiPlatform\Core\Api\IriConverterInterface'

    App\ApiPlatform\AutoGroupResourceMetadataFactory:
        decorates: 'api_platform.metadata.resource.metadata_factory'
        # causes this to decorate around the cached factory so that
        # our service is never cached (which, of course, can have performance
        # implications!
        decoration_priority: -20
        arguments: ['@App\ApiPlatform\AutoGroupResourceMetadataFactory.inner']
```

```yaml
# running bin\console a:o:e -y -o openapi.yaml
# I've removed most of my routes for brevity, but they're all formatted similarly
openapi: 3.0.3
info:
  title: 'Got/IT All Api'
  description: 'The Got/IT Api for Contact Center Management.'
  termsOfService: null
  contact: null
  license: null
  version: 0.1.0
  extensionProperties: []
servers:
  -
    url: /
    description: ''
    variables: null
    extensionProperties: []
paths:
  paths:
    '/api/users/{id}':
      ref: null
      summary: null
      description: null
      get:
        operationId: getUserItem
        tags:
          - user
        responses:
          200:
            description: 'user resource'
            content:
              application/ld+json:
                schema:
                  $ref: '#/components/schemas/user.jsonld-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/html:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/hal+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/csv:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/vnd.api+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.get'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
            headers: null
            links: null
            extensionProperties: []
          404:
            description: 'Resource not found'
            content: null
            headers: null
            links: null
            extensionProperties: []
        summary: 'Retrieves a user resource.'
        description: 'Retrieves a user resource.'
        externalDocs: null
        parameters:
          -
            name: id
            in: path
            description: 'Resource identifier'
            required: true
            deprecated: false
            schema:
              type: string
            style: simple
            example: null
            examples: null
            content: null
            extensionProperties: []
        requestBody: null
        callbacks: null
        deprecated: false
        security: null
        servers: null
        extensionProperties: []
      put:
        operationId: putUserItem
        tags:
          - user
        responses:
          200:
            description: 'user resource updated'
            content:
              application/ld+json:
                schema:
                  $ref: '#/components/schemas/user.jsonld-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/html:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/hal+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/csv:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/vnd.api+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.item.read_user.item.put'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
            headers: null
            links:
              GetUserItem:
                operationId: getUserItem
                parameters:
                  id: '$response.body#/id'
                requestBody: []
                description: 'The `id` value returned in the response can be used as the `id` parameter in `GET /api/users/{id}`.'
                server: null
                extensionProperties: []
            extensionProperties: []
          400:
            description: 'Invalid input'
            content: null
            headers: null
            links: null
            extensionProperties: []
          404:
            description: 'Resource not found'
            content: null
            headers: null
            links: null
            extensionProperties: []
        summary: 'Replaces the user resource.'
        description: 'Replaces the user resource.'
        externalDocs: null
        parameters:
          -
            name: id
            in: path
            description: 'Resource identifier'
            required: true
            deprecated: false
            schema:
              type: string
            style: simple
            example: null
            examples: null
            content: null
            extensionProperties: []
        requestBody:
          description: 'The updated user resource'
          content:
            application/ld+json:
              schema:
                $ref: '#/components/schemas/user.jsonld-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            text/html:
              schema:
                $ref: '#/components/schemas/user-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/hal+json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            text/csv:
              schema:
                $ref: '#/components/schemas/user-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/vnd.api+json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.item.write_user.item.put'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
          required: true
          extensionProperties: []
        callbacks: null
        deprecated: false
        security: null
        servers: null
        extensionProperties: []
      post: null
      delete: null
      options: null
      head: null
      patch: null
      trace: null
      servers: null
      parameters: []
      extensionProperties: []
    /api/users:
      ref: null
      summary: null
      description: null
      get:
        operationId: getUserCollection
        tags:
          - user
        responses:
          200:
            description: 'user collection'
            content:
              application/ld+json:
                schema:
                  type: object
                  properties: { 'hydra:member': { type: array, items: { $ref: '#/components/schemas/user.jsonld-user.read_user.collection.read_user.collection.get' } }, 'hydra:totalItems': { type: integer, minimum: 0 }, 'hydra:view': { type: object, properties: { '@id': { type: string, format: iri-reference }, '@type': { type: string }, 'hydra:first': { type: string, format: iri-reference }, 'hydra:last': { type: string, format: iri-reference }, 'hydra:next': { type: string, format: iri-reference } } }, 'hydra:search': { type: object, properties: { '@type': { type: string }, 'hydra:template': { type: string }, 'hydra:variableRepresentation': { type: string }, 'hydra:mapping': { type: array, items: { type: object, properties: { '@type': { type: string }, variable: { type: string }, property: { type: string, nullable: true }, required: { type: boolean } } } } } } }
                  required: ['hydra:member']
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/json:
                schema:
                  type: array
                  items: { $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.get' }
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/html:
                schema:
                  type: array
                  items: { $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.get' }
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/hal+json:
                schema:
                  type: array
                  items: { $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.get' }
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/csv:
                schema:
                  type: array
                  items: { $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.get' }
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/vnd.api+json:
                schema:
                  type: array
                  items: { $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.get' }
                example: null
                examples: null
                encoding: null
                extensionProperties: []
            headers: null
            links: null
            extensionProperties: []
        summary: 'Retrieves the collection of user resources.'
        description: 'Retrieves the collection of user resources.'
        externalDocs: null
        parameters:
          -
            name: page
            in: query
            description: 'The collection page number'
            required: false
            deprecated: false
            schema:
              type: integer
              default: 1
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: firstName
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: lastName
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: email
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: forgottenPasswordCode
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: 'forgottenPasswordCode[]'
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: array
              items:
                type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: employee.companyEmployeeID
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
          -
            name: 'employee.companyEmployeeID[]'
            in: query
            description: ''
            required: false
            deprecated: false
            schema:
              type: array
              items:
                type: string
            style: form
            example: null
            examples: null
            content: null
            extensionProperties: []
        requestBody: null
        callbacks: null
        deprecated: false
        security: null
        servers: null
        extensionProperties: []
      put: null
      post:
        operationId: postUserCollection
        tags:
          - user
        responses:
          201:
            description: 'user resource created'
            content:
              application/ld+json:
                schema:
                  $ref: '#/components/schemas/user.jsonld-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/html:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/hal+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              text/csv:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
              application/vnd.api+json:
                schema:
                  $ref: '#/components/schemas/user-user.read_user.collection.read_user.collection.post'
                example: null
                examples: null
                encoding: null
                extensionProperties: []
            headers: null
            links:
              GetUserItem:
                operationId: getUserItem
                parameters:
                  id: '$response.body#/id'
                requestBody: []
                description: 'The `id` value returned in the response can be used as the `id` parameter in `GET /api/users/{id}`.'
                server: null
                extensionProperties: []
            extensionProperties: []
          400:
            description: 'Invalid input'
            content: null
            headers: null
            links: null
            extensionProperties: []
        summary: 'Creates a user resource.'
        description: 'Creates a user resource.'
        externalDocs: null
        parameters: []
        requestBody:
          description: 'The new user resource'
          content:
            application/ld+json:
              schema:
                $ref: '#/components/schemas/user.jsonld-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            text/html:
              schema:
                $ref: '#/components/schemas/user-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/hal+json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            text/csv:
              schema:
                $ref: '#/components/schemas/user-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
            application/vnd.api+json:
              schema:
                $ref: '#/components/schemas/user-user.write_user.collection.write_user.collection.post'
              example: null
              examples: null
              encoding: null
              extensionProperties: []
          required: true
          extensionProperties: []
        callbacks: null
        deprecated: false
        security: null
        servers: null
        extensionProperties: []
      delete: null
      options: null
      head: null
      patch: null
      trace: null
      servers: null
      parameters: []
      extensionProperties: []
components:
  schemas:
    user.jsonld-user.read_user.item.read_user.item.get:
      type: object
      description: GotitUser
      properties:
        '@context':
          readOnly: true
          type: string
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user-user.read_user.item.read_user.item.get:
      type: object
      description: GotitUser
      properties:
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user.jsonld-user.read_user.item.read_user.item.put:
      type: object
      description: GotitUser
      properties:
        '@context':
          readOnly: true
          type: string
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user-user.read_user.item.read_user.item.put:
      type: object
      description: GotitUser
      properties:
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user.jsonld-user.write_user.item.write_user.item.put:
      type: object
      description: GotitUser
      properties:
        '@context':
          readOnly: true
          type: string
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
    user-user.write_user.item.write_user.item.put:
      type: object
      description: GotitUser
    user.jsonld-user.read_user.collection.read_user.collection.post:
      type: object
      description: GotitUser
      properties:
        '@context':
          readOnly: true
          type: string
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user-user.read_user.collection.read_user.collection.post:
      type: object
      description: GotitUser
      properties:
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user.jsonld-user.write_user.collection.write_user.collection.post:
      type: object
      description: GotitUser
      properties:
        '@context':
          readOnly: true
          type: string
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
    user-user.write_user.collection.write_user.collection.post:
      type: object
      description: GotitUser
    user.jsonld-user.read_user.collection.read_user.collection.get:
      type: object
      description: GotitUser
      properties:
        '@id':
          readOnly: true
          type: string
        '@type':
          readOnly: true
          type: string
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
    user-user.read_user.collection.read_user.collection.get:
      type: object
      description: GotitUser
      properties:
        id:
          readOnly: true
          type: integer
        shortName:
          readOnly: true
          type: string
          nullable: true
        fullName:
          readOnly: true
          type: string
          nullable: true
  responses: []
  parameters: []
  examples: []
  requestBodies: []
  headers: []
  securitySchemes: []
  links: null
  callbacks: null
  extensionProperties: []
security: []
tags: []
externalDocs: null
extensionProperties: []
```

**Comment by frost-byte** [comment]
I've located the source of my problem.  I'm developing a project remotely using vscode+remote with ssh, I had struggled in the past getting xdebug to work, but was finally able to get the correct configuration for this to work.

This helped me locate a normalizer class that I had created in the past, but had forgotten and did not realize was still being loaded.
Long story short, I found that this normalizer was superseding the one normally used for OpenApi. Once I disabled it, by renaming the file extension, the api docs appeared.

Going to close this now.
